File: compact.py
import copy
from pprint import pprint

# https://people.eecs.berkeley.edu/~keutzer/classes/244fa2005/lectures/5-1-compaction.pdf
# https://blog.finxter.com/5-best-ways-to-find-the-length-of-the-longest-path-in-a-dag-without-repeated-nodes-in-python/
# https://class.ece.uw.edu/541/hauck/lectures/11_Compaction.pdf

def sort_rects(rects, ids, keyfn, reverse=False):
    # from list of rects, select ids and sort 
    # using keyfn ascending or descending
    # return sorted ids
    chosen_rects = [r for r in rects if r['id'] in ids]
    sorted_rects = sorted(chosen_rects, key=keyfn, reverse=reverse)
    sorted_ids = [r['id'] for r in sorted_rects]
    return sorted_ids

def edge_aligned(src, dst, line):
    return \
        src in line['top'] and dst in line['bot'] or \
        src in line['bot'] and dst in line['top']

def compose_graph(lines, rects, dim, reverse=False):
    # lines comes from scanlines_[xy][leftright](...)
    # The graph represents distances from the reference
    # So it doesn't know about rightmost, etc.
    # The reverse flag just adds the dst width/height
    graph = {}
    if   dim == 'x': wh = lambda rect: rect['size'].width()
    elif dim == 'y': wh = lambda rect: rect['size'].height()
    for line in lines:
        src = 0
        for dst in line['sorted']:
            if dst in line['bot']:
                continue
            if (ge := (src, dst)) not in graph:
                if not reverse:
                    graph[ge] = src in rects and wh(rects[src]) or 0
                else:
                    graph[ge] = dst in rects and wh(rects[dst]) or 0
            src = dst
        src = 0
        for dst in line['sorted']:
            if dst in line['top']:
                continue
            if (ge := (src, dst)) not in graph:
                if not reverse:
                    graph[ge] = src in rects and wh(rects[src]) or 0
                else:
                    graph[ge] = dst in rects and wh(rects[dst]) or 0
            src = dst

    return graph                    

def scanlines(rects, axis, reverse=False):
    # rects is list of rectangles
    # Returns list of dicts with edge information
    # Each dict has position and top, mid, bot lists of rect IDs
    # that match that position
    # [{pos:..., 'top':[...], 'mid':[...], 'bot':[...]}, ...]

    #                              top   mid    bot     graph
    #  5  ┌─10┐                     1                   0->1
    # 10  │   │   ┌──15─┐  ┌──15─┐  2,4   1             0->1, 1->2, 2->4
    # 15  │   │   │     │  │  4  │             
    # 20  │   │   │  2  │  ├──15─┤  3    1,2     4      0->1, 1->2, 2->3, 2->4
    # 25  │ 1 │   │     │  │     │         
    # 30  │   │   └─────┘  │  3  │       1,3     2      0->1, 1->2, 2->3
    # 35  │   │            │     │  
    # 40  │   │            └─────┘        1      3      0->1, 1->3
    # 45  └───┘                                  1      0->1
    # 50          ┌─15─┐            5                   0->5
    # 55          │ 5  │     
    # 60          └────┘                         5      0->5

    if axis == 'x':
        prim_pos = lambda rect: rect['pos'].x()
        prim_sz  = lambda rect: rect['size'].width()
        sec_pos  = lambda rect: rect['pos'].y()
        sec_sz   = lambda rect: rect['size'].height()
    elif axis == 'y':
        prim_pos = lambda rect: rect['pos'].y()
        prim_sz  = lambda rect: rect['size'].height()
        sec_pos  = lambda rect: rect['pos'].x()
        sec_sz   = lambda rect: rect['size'].width()

    top_edges = [{'id':r['id'], 'pos':sec_pos(r),             'tb':'top'} for r in rects]
    bot_edges = [{'id':r['id'], 'pos':sec_pos(r) + sec_sz(r), 'tb':'bot'} for r in rects]
    edges = sorted(top_edges + bot_edges, key=lambda edge: edge['pos'])

    # Prime everything with first edge
    edge = edges[0]
    line = {'pos': edge['pos'], 'top':[], 'mid':[], 'bot':[], 'sorted':[edge['id']]}
    line[edge['tb']] = [edge['id']]
    lastpos = edge['pos']

    # Go through each edge.
    # Push accumulated line when next line detected
    lines = []
    for edge in edges[1:]:
        pos = edge['pos']
        id  = edge['id']
        typ = edge['tb']
        if pos > lastpos: # down one edge
            lines.append(copy.deepcopy(line))
            line['pos'] = pos
            line['bot'].clear()
            line['mid'].extend(line['top'])
            line['top'].clear()

        if typ == 'bot' and id in line['mid']:
            line['mid'].remove(id)
        line[typ].append(id)

        line['sorted'] = line['top'] + line['mid'] + line['bot']
        if len(line['sorted']) > 1:
            srtd = sort_rects(rects, line['sorted'], keyfn=prim_pos, reverse=reverse)
            line['sorted'][:] = srtd[:]
        if len(line['top']) > 1:
            srtd = sort_rects(rects, line['top'], keyfn=prim_pos, reverse=reverse)
            line['top'][:] = srtd[:]
        if len(line['bot']) > 1:
            srtd = sort_rects(rects, line['bot'], keyfn=prim_pos, reverse=reverse)
            line['bot'][:] = srtd[:]
        lastpos = pos
    lines.append(copy.deepcopy(line))
    return lines



def make_graph(rects, axis, reverse=False):
    # Returns DAG = {(from_node, to_node): weight, ...}
    # rects is db of rectangles
    # axis is either 'x' or 'y'
    # reverse puts blocks to the right (if 'x') or bottom (if 'y')
    lines = scanlines(rects, axis, reverse)
    graph = compose_graph(lines, rects, axis, reverse)
    return graph

def longest_path_bellman_ford(graph):
    MINDIST = 0
    nodes = set()
    for edge in graph.keys():
        nodes.add(edge[0])
        nodes.add(edge[1])
    num_nodes = len(nodes)
    dist = [float('-inf')] * num_nodes
    dist[0] = 0  # Assuming 0 is the start node

    # looks like O(num_nodes * num_edges)
    for _ in range(num_nodes - 1):
        for (frm, to), weight in graph.items():
            weight += MINDIST
            if dist[frm] != float('-inf') and dist[frm] + weight > dist[to]:
                dist[to] = dist[frm] + weight
        return dist

def randgraph(numnodes, numedges, weightrng):
    # Generate random graph
    from random import randint
    rnd_node = lambda: randint(0, numnodes - 1)
    rnd_weight = lambda: randint(0, weightrng)
    rnd_srcdst = lambda: randint(0, 1)
    graph = {}
    for n in range(numnodes):
        other = rnd_node()
        while n == other:
            other = rnd_node()
        if rnd_srcdst():
            graph[(n, other)] = rnd_weight()
        else:
            graph[(other, n)] = rnd_weight()
    for _ in range(numedges - numnodes):
        frm = rnd_node()
        to = rnd_node()
        while frm == to or (frm, to) in graph:
            frm = rnd_node()
            to = rnd_node()
        graph[(frm, to)] = rnd_weight()
    return graph

if __name__ == '__main__':
    # graph = randgraph(5, 10, 20)
    # pprint(graph)

    import timeit
    num = 1000
    loop = 5
    numnodes = 1000
    numedges = 4*numnodes
    weight_rng = 50
    runstr1 = f'graph = randgraph(numnodes, numedges, weight_rng)'
    result = timeit.timeit(runstr1, globals=globals(), number=num)
    print(f'Create graph time per loop is {1000 * result/num:.5f} ms')

    graph = randgraph(numnodes, numedges, weight_rng)
    runstr2 = 'lp = longest_path_bellman_ford(graph)'
    result = timeit.timeit(runstr2, globals=globals(), number=num)
    print(f'Longest path time per loop is {1000 * result/num:.5f} ms')
    

File: compact.nim
import std/[algorithm, locks, sugar, tables]
import sequtils
import wnim
import wnim/wTypes
import winim/inc/winuser
import concurrent
import recttable, userMessages

type 
  Axis* = enum X=true, Y=false
  MajMin = enum Major=true, Minor=false
  Node = RectID
  Weight = int
  GraphEdge  = tuple[frm, to: Node]
  Graph* = Table[GraphEdge, Weight]
  ScanType = enum Top, Mid, Bot
  ScanEdge = tuple
    id:    RectID
    pos:   int
    etype: ScanType
  ScanLine = tuple
    pos:    int        
    top:    seq[RectID]
    mid:    seq[RectID]
    bot:    seq[RectID]
    sorted: seq[RectID]
  CompactDir* = tuple
    primax,  secax:  Axis
    primAsc, secAsc: SortOrder
  CompoundDir* = enum UpLeft,UpRight,DownLeft,DownRight,LeftUp,LeftDown,RightUp,RightDown
  CompactArg* = tuple
    pRectTable: ptr RectTable
    direction:  CompactDir
    window:     wWindow
    dstRect:    wRect


const
  RootNode = when RectID is string: "0" else: 0

var
  gCompactThread*: Thread[CompactArg]

proc compoundDir*(cd: CompactDir): CompoundDir =
  let compound = (cd.primax == X, cd.primAsc, cd.secAsc)
  if   compound == (false, Ascending,  Ascending ): UpLeft
  elif compound == (false, Ascending,  Descending): UpRight
  elif compound == (false, Descending, Ascending ): DownLeft
  elif compound == (false, Descending, Descending): DownRight
  elif compound == (true,  Ascending,  Ascending ): LeftUp
  elif compound == (true,  Ascending,  Descending): LeftDown
  elif compound == (true,  Descending, Ascending ): RightUp
  else: RightDown
  
proc isXAscending*(direction: CompactDir): bool =
  (direction.primax == X and direction.primAsc == Ascending) or
  (direction.secax  == X and direction.secAsc  == Ascending)

proc isYAscending*(direction: CompactDir): bool = 
  (direction.primax == Y and direction.primAsc == Ascending) or
  (direction.secax  == Y and direction.secAsc  == Ascending)

proc rectCmpX(r1, r2: Rect): int = 
  # Sort first by x position, then by id
  result = cmp(r1.x, r2.x)
  if result == 0:
    result = cmp(r1.id, r2.id)

proc rectCmpY(r1, r2: Rect): int = 
  # Sort first by y position, then by id
  result = cmp(r1.y, r2.y)
  if result == 0:
    result = cmp(r1.id, r2.id)

# TODO: change to in-place sorting
proc sortedRectsIds(rects: seq[Rect], axis: Axis, sortOrder: SortOrder): seq[RectID] =
  # Returns rect ids with compare chosen by axis
  var tmpRects = rects
  if axis == X:
    tmpRects.sort(rectCmpX, sortOrder)
  else:
    tmpRects.sort(rectCmpY, sortOrder)
  result = tmpRects.ids



proc setField[T](line: var ScanLine, field: ScanType, val: T) =
  case field
    of Top: line.top = val
    of Mid: line.mid = val
    of Bot: line.bot = val

proc appendField[T](line: var ScanLine, field: ScanType, val: T) =
  case field
    of Top: line.top.add(val)
    of Mid: line.mid.add(val)
    of Bot: line.bot.add(val)

type DimGetter = proc(node: Node): int
proc MakeDimGetter(rectTable: RectTable, axis: Axis): DimGetter =
  if axis == X:
    proc(node: Node): int =
      if node != RootNode:
        result = rectTable[node].wRect.width
  else: # axis == Y:
    proc(node: Node): int =
      if node != RootNode:
        result = rectTable[node].wRect.height

proc ComposeGraph(lines: seq[ScanLine], rectTable: RectTable,
                  axis: Axis, sortOrder: SortOrder): Graph = 
  let getDim = MakeDimGetter(rectTable, axis)
  var src: Node
  for line in lines:
    src = RootNode
    for dst in line.sorted:
      if dst in line.bot: continue
      if (src, dst) notin result:
        result[(src, dst)] = if sortOrder == Descending: dst.getDim else: src.getDim
      src = dst
    src = RootNode
    for dst in line.sorted:
      if dst in line.top: continue
      if (src, dst) notin result:
        result[(src, dst)] = if sortOrder == Descending: dst.getDim else: src.getDim
      src = dst

proc PosChooser(ax: MajMin): proc(rect: Rect): int =
  if ax == Major:
    proc(rect: Rect): int =  rect.x
  else:
    proc(rect: Rect): int =  rect.y

proc SizeChooser(ax: MajMin): proc(rect: Rect): int =
  if ax == Major:
    proc(rect: Rect): int = rect.wRect.width
  else:
    proc(rect: Rect): int = rect.wRect.height

proc ScanLines(rectTable: RectTable, axis: Axis, sortOrder: SortOrder, ids: seq[RectID]): seq[ScanLine] =
  let 
    minor = if axis==X: Minor else: Major
    SecPos  = PosChooser(minor)
    SecSz   = SizeChooser(minor)
    colids  = if ids.len == 0: rectTable.keys.toSeq
             else:             ids
    topEdges: seq[ScanEdge] = 
        collect(for id in colids:
          (id: id, pos: rectTable[id].SecPos, etype: Top))
    botEdges: seq[ScanEdge] = 
        collect (for id in colids:
          let rect = rectTable[id]
          (id: id, pos: rect.SecPos + rect.SecSz, etype: Bot))
    edges: seq[ScanEdge] = concat(topEdges, botEdges).sortedByIt(it.pos)
  
  # Prime everything with first edge
  var edge: ScanEdge  = edges[0]
  var line: ScanLine  = (pos: edge.pos, 
                         top: @[], mid: @[], bot: @[], 
                         sorted: @[edge.id])
  line.setField(edge.etype, @[edge.id])
  var lastpos = edge.pos

  # Go through each edge
  # Push accumulated line when next line detected
  for i, edge in edges[1..high(edges)]:
    if edge.pos > lastpos: # down one edge
      result.add(line)
      line.pos = edge.pos
      line.bot = @[] # TODO: profile vs. setLen(0)
      line.mid.add(line.top)
      line.top = @[]

    if edge.etype == Bot and edge.id in line.mid:
      line.mid.delete(line.mid.find(edge.id))
    line.appendField(edge.etype, edge.id)

    line.sorted = concat(line.top, line.mid, line.bot)
    if line.sorted.len > 1:
      line.sorted = sortedRectsIds(rectTable[line.sorted], axis, sortOrder)
    if line.top.len > 1:
      line.top = sortedRectsIds(rectTable[line.top], axis, sortOrder)
    if line.bot.len > 1:
      line.bot = sortedRectsIds(rectTable[line.bot], axis, sortOrder)
    lastpos = edge.pos
  result.add(line)

proc MakeGraph*(rectTable: RectTable, axis: Axis, sortOrder: SortOrder, ids: seq[RectID]): Graph =
  # Returns DAG = table((frm,to): weight)
  # rectTable is table of rects
  # axis is X or Y
  # sortOrder == left/up or down/right
  let lines = ScanLines(rectTable, axis, sortOrder, ids)
  result = ComposeGraph(lines, rectTable, axis, sortOrder)

proc longestPathBellmanFord(graph: Graph, nodes: openArray[Node], minpos: int): Table[RectID, Weight] =
  for node in nodes:
    result[node] = Weight.low
  result[RootNode] = minpos

  for iter in 0..nodes.len:
    for ge, weight in graph:
      if result[ge.frm] == Weight.low:              
        continue
      result[ge.to] = max(result[ge.to], result[ge.frm] + weight)
  result.del(RootNode)

proc compact*(rectTable: RectTable, 
              axis: Axis,
              sortOrder: SortOrder,
              dstRect: wRect,
              ids: seq[RectID] = @[]) =
  # Top level compact function in one direction
  let graph = MakeGraph(rectTable, axis, sortOrder, ids)
  let nodes = if ids.len == 0: rectTable.keys.toSeq
              else: ids

  if axis == X and sortOrder == Ascending:
    let lp = longestPathBellmanFord(graph, nodes, 0)
    for id in nodes:
      rectTable[id].x = dstRect.x + lp[id]

  elif axis == X and sortOrder == Descending:
    let lp = longestPathBellmanFord(graph, nodes, 0)
    for id in nodes:
      rectTable[id].x = dstRect.x + dstRect.width - lp[id]

  elif axis == Y and sortOrder == Ascending:
    let lp = longestPathBellmanFord(graph, nodes, 0)
    for id in nodes:
      rectTable[id].y = dstRect.y + lp[id]

  elif axis == Y and sortOrder == Descending:
    let lp = longestPathBellmanFord(graph, nodes, 0)
    for id in nodes:
      rectTable[id].y = dstRect.y + dstRect.height - lp[id]

proc iterCompact*(rectTable: RectTable, direction: CompactDir, dstRect: wRect) =
  # Run compact function until rectTable doesn't change
  var pos, lastPos: PosTable
  pos = rectTable.positions
  while pos != lastPos:
    compact(rectTable, direction.primax, direction.primAsc, dstRect)
    compact(rectTable, direction.secax, direction.secAsc, dstRect)
    lastPos = pos
    pos = rectTable.positions

proc compactWorker*(arg: CompactArg) {.thread.} =
  {.gcsafe.}:
    withLock(gLock):
      iterCompact(arg.pRectTable[], arg.direction, arg.dstRect)
  PostMessage(arg.window.mHwnd, USER_ALG_UPDATE, 0, 0)
  